clusterd: ensure system vars apply to persist PubSub connection

benesch opened the pull request:
Keeps `clusterd` from establishing its persist PubSub
connection until `environmentd` sends over its initial batch of
configuration values. This ensures that any parameters that should apply
to the connection (e.g., timeouts) actually take effect.

This relies on dyncfg's newly-added support for waiting for a
configuration update to be applied.

Fix https://github.com/MaterializeInc/materialize/issues/23869.

### Motivation

  * This PR fixes a recognized bug.

### Tips for reviewer

@danhhz: there are some potentially 🌶️ changes to dyncfg in the earlier commits. Might be helpful to discuss on Slack or live before you get too deep.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

benesch commented on src/dyncfg/src/lib.rs:
```diff
@@ -217,6 +206,64 @@ impl ConfigSet {
+    ///
+    /// `f` should take care to execute quickly, as no further configuration
+    /// updates can be applied while `f` is executing.
+    pub fn with_updates_frozen<F>(&self, mut f: F)
```
This isn't presently used, but it seemed like it could be useful in the future!

guswynn replied:
it DOES seem useful, for some storage stuff

danhhz commented on src/dyncfg/src/lib.rs:
```diff
@@ -119,19 +119,6 @@ impl<T: ConfigType> Config<T> {
 
-    /// Adds this config to the given set.
-    ///
-    /// Names are required to be unique within a set, but each set is entirely
```
Wild copy-pasta here, presumably by me. Let's move this full rustdoc to [ConfigSet::add] so we don't lose the bit of additional detail

benesch replied:
Done.

danhhz commented on src/dyncfg/src/lib.rs:
```diff
@@ -234,44 +221,118 @@ impl ConfigEntry {
+            (ConfigValAtomic::Duration(x), ConfigVal::Duration(val)) => {
+                *x.write().expect("lock poisoned") = val
+            }
+            (_, val) => panic!("attempted to store {val:?} value in {self:?} parameter"),
```
Oh lol. I mentioned in slack that I loved us getting rid of the "Restructure this match so that it's harder to miss when adding a new `ConfigType`." TODO, but I think we kept the problem and just lost the TODO :)

At least the single way to set/get things means we have an easier time making sure it's covered by unit tests, so I still think this is an improvement

benesch replied:
Oh, it's easy enough to make this exhaustive. Done!

danhhz commented on src/dyncfg/src/lib.rs:
```diff
@@ -234,44 +221,118 @@ impl ConfigEntry {
     /// If a value of the same config has previously been added to these
     /// updates, replaces it.
-    pub fn add(&mut self, config: &ConfigEntry) {
+    pub fn add<T>(&mut self, config: &Config<T>, val: T)
```
nit: when inlining the `where` in the type params doesn't cause it to wrap, I tend to do it (for information density reasons). i.e. `pub fn add<T: ConfigType>(&mut self, config: &Config<T>, val: T) {`

Here and add_dynamic

danhhz commented on src/dyncfg/src/lib.rs:
```diff
@@ -225,6 +221,43 @@ impl ConfigEntry {
+where
+    T: ConfigType,
+{
+    #[ref_cast_custom]
```
I don't love the ref_cast stuff, lots of magic, macros, and third party crates.

Would something like this work instead: keep `Config::shared` but make it not-pub, keep `Config::get` implemented in terms of `shared`, in `Config::handle` clone the `ConfigValAtomic` and use it to construct the handle?

benesch replied:
Can do, done.

danhhz reviewed (commented):
As discussed in slack, not sold yet on commit 7, but reviewed 1-6 because they seem good to have and we could pull them out into another PR while we hash out 7

guswynn reviewed (commented):
non-dyncfg stuff lgtm, thanks for closing this issue! Ill leave the accept to @danhhz !

benesch commented:
@danhhz per our conversation a few weeks ago:
* I've addressed feedback, and split out the first couple commits into https://github.com/MaterializeInc/materialize/pull/26347, which is, as far as I understand, basically ready to merge!
* I'm going to leave this PR up as a reference implementation for the `wait_for` approach. 
* I'm going to put up a separate PR that has the same effect as this PR but uses a `has_synced_once` bool in the `PersistConfig`, like we discussed, and we can compare to see which looks better!
